File: src/main/java/com/douradelivery/service/DocumentService.java
package com.douradelivery.service;

import com.douradelivery.model.Document;
import com.douradelivery.model.Document.DocumentStatus;
import com.douradelivery.model.Document.DocumentType;
import com.douradelivery.model.User;
import com.douradelivery.repository.DocumentRepository;
import com.douradelivery.repository.UserRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class DocumentService {
    
    private static final Logger logger = LoggerFactory.getLogger(DocumentService.class);
    
    @Autowired
    private DocumentRepository documentRepository;
    
    @Autowired
    private UserRepository userRepository;
    
    @Value("${app.upload.dir:uploads}")
    private String uploadDir;
    
    public Map<String, Object> uploadDocuments(Long userId, Map<String, MultipartFile> files) {
        try {
            logger.info("Iniciando upload de documentos para usuário: {}", userId);
            
            Optional<User> userOpt = userRepository.findById(userId);
            if (!userOpt.isPresent()) {
                return createErrorResponse("Usuário não encontrado");
            }
            
            User user = userOpt.get();
            
            // Criar diretório de upload se não existir
            Path uploadPath = Paths.get(uploadDir);
            if (!Files.exists(uploadPath)) {
                Files.createDirectories(uploadPath);
            }
            
            List<Document> documents = new ArrayList<>();
            
            // Processar cada arquivo
            for (Map.Entry<String, MultipartFile> entry : files.entrySet()) {
                String fieldName = entry.getKey();
                MultipartFile file = entry.getValue();
                
                if (file != null && !file.isEmpty()) {
                    DocumentType documentType = mapFieldToDocumentType(fieldName);
                    if (documentType != null) {
                        Document document = saveFile(user, file, documentType);
                        if (document != null) {
                            documents.add(document);
                        }
                    }
                }
            }
            
            if (documents.isEmpty()) {
                return createErrorResponse("Nenhum documento válido foi enviado");
            }
            
            // Salvar documentos no banco
            documentRepository.saveAll(documents);
            
            logger.info("Upload concluído com sucesso. {} documentos salvos para usuário {}", 
                       documents.size(), userId);
            
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", "Documentos enviados com sucesso!");
            response.put("documentsCount", documents.size());
            
            return response;
            
        } catch (Exception e) {
            logger.error("Erro ao fazer upload de documentos para usuário {}: {}", userId, e.getMessage(), e);
            return createErrorResponse("Erro interno do servidor: " + e.getMessage());
        }
    }
    
    private Document saveFile(User user, MultipartFile file, DocumentType documentType) {
        try {
            // Gerar nome único para o arquivo
            String originalFilename = file.getOriginalFilename();
            String extension = originalFilename != null && originalFilename.contains(".") 
                ? originalFilename.substring(originalFilename.lastIndexOf("."))
                : ".jpg";
            
            String fileName = String.format("%s_%d_%d%s", 
                documentType.name().toLowerCase(),
                user.getId(),
                System.currentTimeMillis(),
                extension);
            
            // Salvar arquivo no sistema de arquivos
            Path filePath = Paths.get(uploadDir, fileName);
            Files.write(filePath, file.getBytes());
            
            // Criar registro no banco
            Document document = new Document(user, documentType, fileName, filePath.toString());
            
            logger.info("Arquivo salvo: {} para usuário {}", fileName, user.getId());
            
            return document;
            
        } catch (IOException e) {
            logger.error("Erro ao salvar arquivo para usuário {}: {}", user.getId(), e.getMessage(), e);
            return null;
        }
    }
    
    private DocumentType mapFieldToDocumentType(String fieldName) {
        switch (fieldName.toLowerCase()) {
            case "cpfdocument":
                return DocumentType.CPF;
            case "profilephoto":
                return DocumentType.PROFILE;
            case "cnhdocument":
                return DocumentType.CNH;
            case "vehicledocument":
                return DocumentType.VEHICLE;
            default:
                logger.warn("Tipo de documento não reconhecido: {}", fieldName);
                return null;
        }
    }
    
    public Map<String, Object> getDocumentStatus(Long userId) {
        try {
            logger.info("Buscando status dos documentos para usuário: {}", userId);
            
            Optional<User> userOpt = userRepository.findById(userId);
            if (!userOpt.isPresent()) {
                return createErrorResponse("Usuário não encontrado");
            }
            
            List<Document> documents = documentRepository.findByUserIdOrderBySubmittedAtDesc(userId);
            
            if (documents.isEmpty()) {
                Map<String, Object> response = new HashMap<>();
                response.put("success", true);
                response.put("status", "NOT_SUBMITTED");
                response.put("message", "Nenhum documento enviado");
                return response;
            }
            
            // Pegar o documento mais recente para determinar o status geral
            Document latestDocument = documents.get(0);
            
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("status", latestDocument.getStatus().name());
            response.put("observation", latestDocument.getObservation());
            response.put("submittedAt", latestDocument.getSubmittedAt());
            response.put("reviewedAt", latestDocument.getReviewedAt());
            
            // Adicionar lista de documentos
            List<Map<String, Object>> documentList = documents.stream()
                .map(this::documentToMap)
                .collect(Collectors.toList());
            response.put("documents", documentList);
            
            return response;
            
        } catch (Exception e) {
            logger.error("Erro ao buscar status dos documentos para usuário {}: {}", userId, e.getMessage(), e);
            return createErrorResponse("Erro interno do servidor");
        }
    }
    
    public List<Map<String, Object>> getPendingDocuments() {
        try {
            logger.info("Buscando documentos pendentes para análise");
            
            List<Document> pendingDocuments = documentRepository.findPendingDocumentsWithUser();
            
            // Agrupar por usuário
            Map<Long, List<Document>> documentsByUser = pendingDocuments.stream()
                .collect(Collectors.groupingBy(doc -> doc.getUser().getId()));
            
            List<Map<String, Object>> result = new ArrayList<>();
            
            for (Map.Entry<Long, List<Document>> entry : documentsByUser.entrySet()) {
                List<Document> userDocuments = entry.getValue();
                if (!userDocuments.isEmpty()) {
                    Document firstDoc = userDocuments.get(0);
                    User user = firstDoc.getUser();
                    
                    Map<String, Object> userDocumentMap = new HashMap<>();
                    userDocumentMap.put("userId", user.getId());
                    userDocumentMap.put("userName", user.getName());
                    userDocumentMap.put("userType", user.getUserType().name());
                    userDocumentMap.put("submittedAt", firstDoc.getSubmittedAt());
                    
                    List<Map<String, Object>> documents = userDocuments.stream()
                        .map(this::documentToMap)
                        .collect(Collectors.toList());
                    userDocumentMap.put("documents", documents);
                    
                    result.add(userDocumentMap);
                }
            }
            
            // Ordenar por data de submissão (mais antigos primeiro)
            result.sort((a, b) -> {
                LocalDateTime dateA = (LocalDateTime) a.get("submittedAt");
                LocalDateTime dateB = (LocalDateTime) b.get("submittedAt");
                return dateA.compareTo(dateB);
            });
            
            logger.info("Encontrados {} usuários com documentos pendentes", result.size());
            
            return result;
            
        } catch (Exception e) {
            logger.error("Erro ao buscar documentos pendentes: {}", e.getMessage(), e);
            return new ArrayList<>();
        }
    }
    
    public Map<String, Object> approveDocuments(Long userId, String observation, Long reviewerId) {
        return updateDocumentStatus(userId, DocumentStatus.APPROVED, observation, reviewerId);
    }
    
    public Map<String, Object> rejectDocuments(Long userId, String observation, Long reviewerId) {
        return updateDocumentStatus(userId, DocumentStatus.REJECTED, observation, reviewerId);
    }
    
    private Map<String, Object> updateDocumentStatus(Long userId, DocumentStatus status, 
                                                   String observation, Long reviewerId) {
        try {
            logger.info("Atualizando status dos documentos do usuário {} para {}", userId, status);
            
            Optional<User> userOpt = userRepository.findById(userId);
            if (!userOpt.isPresent()) {
                return createErrorResponse("Usuário não encontrado");
            }
            
            Optional<User> reviewerOpt = userRepository.findById(reviewerId);
            if (!reviewerOpt.isPresent()) {
                return createErrorResponse("Revisor não encontrado");
            }
            
            List<Document> userDocuments = documentRepository.findByUserIdAndStatus(userId, DocumentStatus.PENDING);
            
            if (userDocuments.isEmpty()) {
                return createErrorResponse("Nenhum documento pendente encontrado para este usuário");
            }
            
            // Atualizar todos os documentos pendentes do usuário
            LocalDateTime now = LocalDateTime.now();
            User reviewer = reviewerOpt.get();
            
            for (Document document : userDocuments) {
                document.setStatus(status);
                document.setObservation(observation);
                document.setReviewedAt(now);
                document.setReviewedBy(reviewer);
            }
            
            documentRepository.saveAll(userDocuments);
            
            logger.info("Status dos documentos do usuário {} atualizado para {} por revisor {}", 
                       userId, status, reviewerId);
            
            Map<String, Object> response = new HashMap<>();
            response.put("success", true);
            response.put("message", String.format("Documentos %s com sucesso!", 
                status == DocumentStatus.APPROVED ? "aprovados" : "rejeitados"));
            response.put("status", status.name());
            response.put("documentsUpdated", userDocuments.size());
            
            return response;
            
        } catch (Exception e) {
            logger.error("Erro ao atualizar status dos documentos do usuário {}: {}", userId, e.getMessage(), e);
            return createErrorResponse("Erro interno do servidor");
        }
    }
    
    private Map<String, Object> documentToMap(Document document) {
        Map<String, Object> map = new HashMap<>();
        map.put("id", document.getId());
        map.put("type", document.getDocumentType().name());
        map.put("fileName", document.getFileName());
        map.put("status", document.getStatus().name());
        map.put("submittedAt", document.getSubmittedAt());
        map.put("reviewedAt", document.getReviewedAt());
        map.put("observation", document.getObservation());
        return map;
    }
    
    private Map<String, Object> createErrorResponse(String message) {
        Map<String, Object> response = new HashMap<>();
        response.put("success", false);
        response.put("message", message);
        return response;
    }
}

File: flutter_app/lib/services/document_service.dart
import 'package:dio/dio.dart';
import 'dart:io';
import 'dart:typed_data';
import 'package:flutter/foundation.dart';
import 'api_service.dart';

class DocumentService {
  final ApiService _apiService = ApiService();

  Future<Map<String, dynamic>> uploadDocuments({
    required int userId,
    required File cpfDocument,
    required File profilePhoto,
    File? cnhDocument,
    File? vehicleDocument,
  }) async {
    try {
      print('📤 Enviando documentos para usuário $userId');

      // Verificar se o backend está disponível
      bool backendAvailable = false;
      try {
        await _apiService.dio.get('/actuator/health');
        backendAvailable = true;
        print('✅ Backend disponível - fazendo upload real');
      } catch (e) {
        print('⚠️ Backend indisponível - simulando upload');
        backendAvailable = false;
      }

      if (!backendAvailable) {
        // Simular upload quando backend não está disponível
        print('📱 Simulando upload...');
        await Future.delayed(const Duration(seconds: 2)); // Simular delay de upload
        
        return {
          'success': true,
          'message': 'Documentos enviados com sucesso! (Simulado - Backend em deploy)',
          'documentsCount': 2 + (cnhDocument != null ? 1 : 0) + (vehicleDocument != null ? 1 : 0),
        };
      }

      // Para mobile/desktop, usar upload real
      final formData = FormData.fromMap({
        'userId': userId,
        'cpfDocument': await MultipartFile.fromFile(
          cpfDocument.path,
          filename: 'cpf_${userId}_${DateTime.now().millisecondsSinceEpoch}.jpg',
        ),
        'profilePhoto': await MultipartFile.fromFile(
          profilePhoto.path,
          filename: 'profile_${userId}_${DateTime.now().millisecondsSinceEpoch}.jpg',
        ),
        if (cnhDocument != null)
          'cnhDocument': await MultipartFile.fromFile(
            cnhDocument.path,
            filename: 'cnh_${userId}_${DateTime.now().millisecondsSinceEpoch}.jpg',
          ),
        if (vehicleDocument != null)
          'vehicleDocument': await MultipartFile.fromFile(
            vehicleDocument.path,
            filename: 'vehicle_${userId}_${DateTime.now().millisecondsSinceEpoch}.jpg',
          ),
      });

      final response = await _apiService.dio.post(
        '/api/documents/upload',
        data: formData,
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        print('✅ Documentos enviados com sucesso!');
        return {
          'success': true,
          'message': 'Documentos enviados com sucesso!',
          ...response.data,
        };
      } else {
        throw Exception('Erro ao enviar documentos: Status ${response.statusCode}');
      }
    } on DioException catch (e) {
      print('❌ Erro ao enviar documentos:');
      print('   Status: ${e.response?.statusCode}');
      print('   Response: ${e.response?.data}');

      if (e.response != null) {
        final statusCode = e.response!.statusCode;
        final responseData = e.response!.data;

        String errorMessage = 'Erro desconhecido';

        if (responseData is Map<String, dynamic> && responseData.containsKey('message')) {
          errorMessage = responseData['message'];
        } else {
          switch (statusCode) {
            case 400:
              errorMessage = 'Dados inválidos. Verifique os documentos enviados.';
              break;
            case 413:
              errorMessage = 'Arquivos muito grandes. Reduza o tamanho das imagens.';
              break;
            case 500:
              errorMessage = 'Erro interno do servidor. Tente novamente.';
              break;
            default:
              errorMessage = 'Erro no servidor (Status $statusCode)';
          }
        }

        return {'success': false, 'message': errorMessage};
      } else {
        return {'success': false, 'message': 'Erro de conexão: ${e.message}'};
      }
    } catch (e) {
      print('❌ Erro inesperado: $e');
      return {'success': false, 'message': 'Erro inesperado: $e'};
    }
  }

  Future<Map<String, dynamic>> getDocumentStatus(int userId) async {
    try {
      print('📥 Buscando status dos documentos para usuário $userId');

      final response = await _apiService.dio.get('/api/documents/status/$userId');

      if (response.statusCode == 200) {
        print('✅ Status dos documentos obtido com sucesso!');
        return {
          'success': true,
          ...response.data,
        };
      } else {
        throw Exception('Erro ao buscar status: Status ${response.statusCode}');
      }
    } on DioException catch (e) {
      print('❌ Erro ao buscar status dos documentos:');
      print('   Status: ${e.response?.statusCode}');
      print('   Response: ${e.response?.data}');

      return {
        'success': false,
        'message': 'Erro ao buscar status dos documentos',
        'status': 'PENDING',
      };
    } catch (e) {
      print('❌ Erro inesperado: $e');
      return {
        'success': false,
        'message': 'Erro inesperado: $e',
        'status': 'PENDING',
      };
    }
  }

  Future<List<Map<String, dynamic>>> getPendingDocuments() async {
    try {
      print('📥 Buscando documentos pendentes para análise');

      final response = await _apiService.dio.get('/api/admin/documents/pending');

      if (response.statusCode == 200) {
        print('✅ Documentos pendentes obtidos com sucesso!');
        return List<Map<String, dynamic>>.from(response.data);
      } else {
        throw Exception('Erro ao buscar documentos: Status ${response.statusCode}');
      }
    } on DioException catch (e) {
      print('❌ Erro ao buscar documentos pendentes:');
      print('   Status: ${e.response?.statusCode}');
      print('   Response: ${e.response?.data}');
      return [];
    } catch (e) {
      print('❌ Erro inesperado: $e');
      return [];
    }
  }

  Future<Map<String, dynamic>> approveDocument({
    required int userId,
    required String observation,
  }) async {
    try {
      print('✅ Aprovando documentos do usuário $userId');

      final response = await _apiService.dio.post(
        '/api/admin/documents/approve',
        data: {
          'userId': userId,
          'status': 'APPROVED',
          'observation': observation,
        },
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        print('✅ Documentos aprovados com sucesso!');
        return {
          'success': true,
          'message': 'Documentos aprovados com sucesso!',
          ...response.data,
        };
      } else {
        throw Exception('Erro ao aprovar documentos: Status ${response.statusCode}');
      }
    } on DioException catch (e) {
      print('❌ Erro ao aprovar documentos:');
      print('   Status: ${e.response?.statusCode}');
      print('   Response: ${e.response?.data}');

      return {
        'success': false,
        'message': 'Erro ao aprovar documentos',
      };
    } catch (e) {
      print('❌ Erro inesperado: $e');
      return {
        'success': false,
        'message': 'Erro inesperado: $e',
      };
    }
  }

  Future<Map<String, dynamic>> rejectDocument({
    required int userId,
    required String observation,
  }) async {
    try {
      print('❌ Rejeitando documentos do usuário $userId');

      final response = await _apiService.dio.post(
        '/api/admin/documents/reject',
        data: {
          'userId': userId,
          'status': 'REJECTED',
          'observation': observation,
        },
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        print('✅ Documentos rejeitados com sucesso!');
        return {
          'success': true,
          'message': 'Documentos rejeitados com sucesso!',
          ...response.data,
        };
      } else {
        throw Exception('Erro ao rejeitar documentos: Status ${response.statusCode}');
      }
    } on DioException catch (e) {
      print('❌ Erro ao rejeitar documentos:');
      print('   Status: ${e.response?.statusCode}');
      print('   Response: ${e.response?.data}');

      return {
        'success': false,
        'message': 'Erro ao rejeitar documentos',
      };
    } catch (e) {
      print('❌ Erro inesperado: $e');
      return {
        'success': false,
        'message': 'Erro inesperado: $e',
      };
    }
  }
}